benchmarks/benchmark_base.py

```diff
@@ -0,0 +1,453 @@
+
+            Parameters
+            ----------
+            data : ndarray
+                The ndarray to be synchronized.
+
+            Returns
+            -------
+            ndarray or None
+                The synchronized ndarray if `enable_generate_reference` is `False`, otherwise `None`.
+            """
+            self.fname = f"{self.fname_prefix}.npy"
+            if self.enable_generate_reference:
+                Path(self.fpath).parent.mkdir(parents=True, exist_ok=True)
+                Path(self.fpath).parent.mkdir(parents=True, exist_ok=True)
+                np.save(self.fpath, data)
+                raise Exception("Skipping test to generate reference data")
+            else:
+                return np.load(self.fpath)
+
+        def sync_str(self, data):
+            """
+            Synchronizes the string with the regression data.
+
+            Parameters
+            ----------
+            data : str
+                The string to be synchronized.
+
+            Returns
+            -------
+            str or None
+                The synchronized string if `enable_generate_reference` is `False`, otherwise `None`.
+            """
+            self.fname = f"{self.fname_prefix}.txt"
+            if self.enable_generate_reference:
+                Path(self.fpath).parent.mkdir(parents=True, exist_ok=True)
+                with Path(self.fpath).open("w") as fh:
+                    fh.write(data)
+                raise Exception(
+                    f"Skipping test to generate regression_data {self.fpath} data"
+                )
+            else:
+                with Path(self.fpath).open("r") as fh:
+                    return fh.read()
+
+        def sync_hdf_store(self, tardis_module, update_fname=True):
+            """
+            Synchronizes the HDF store with the regression data.
+
+            Parameters
+            ----------
+            tardis_module : object
+                The module to be synchronized.
+            update_fname : bool, optional
+                Whether to update the file name. Defaults to True.
+
+            Returns
+            -------
+            HDFStore or None
+                The synchronized HDF store if `enable_generate_reference` is `False`, otherwise `None`.
+            """
+            if update_fname:
+                self.fname = f"{self.fname_prefix}.h5"
+            if self.enable_generate_reference:
+                Path(self.fpath).parent.mkdir(parents=True, exist_ok=True)
+                with pd.HDFStore(self.fpath, mode="w") as store:
+                    tardis_module.to_hdf(store, overwrite=True)
+                raise Exception(
+                    f"Skipping test to generate regression_data {self.fpath} data"
+                )
+            else:
+                return pd.HDFStore(self.fpath, mode="r")
+
+    @property
+    def regression_data(self):
+        return self.RegressionData()
+
+    @property
+    def packet(self):
```

Comment:
Please do

Reply:
This is my best proposal for the re-utilization Regression Data class in PyTest: ab98e618580e8e1efe39715d231cbba499d09cf9.

The big problem is that PyTest when runs, it got many values and information, and it is passed and used in the Fixtures. It is challenging to get these values running ASV, we need to create fake objects and hard-code some values.

For this demo, I ran ASV with `asv run --bench "time_prepare_bound_bound_rate_matrix" --bench "time_coll_exc_deexc_matrix"`. It works good, basically took the regression data, it did some processes, and compared the results for the process and the regression data.

The output was:

```
· Creating environments
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[ 0.00%] · For tardis commit b7e07720 <master>:
[ 0.00%] ·· Benchmarking mamba-py3.12
[25.00%] ··· Running (plasma_nlte_excitation_x.BenchmarkPlasmaNlteExcitation.time_coll_exc_deexc_matrix--)..
[75.00%] ··· plasma_nlte_excitation_x.BenchmarkPlasmaNlteExcitation.time_coll_exc_deexc_matrix                                                 ok
[75.00%] ··· ============================================================================================================================== =============
                                                                         Matrix
             ------------------------------------------------------------------------------------------------------------------------------ -------------
                          {'coll_exc_coeff_values': [1, -2, 3], 'coll_deexc_coeff_values': [4, 9, 10], 'number_of_levels': 3}                1.60±0.04ms
              {'coll_exc_coeff_values': [0.21, 0.045, 0.1234], 'coll_deexc_coeff_values': [0.7865, 0.987, 0.00123], 'number_of_levels': 3}   1.63±0.04ms
             ============================================================================================================================== =============

[100.00%] ··· plasma_nlte_excitation_x.BenchmarkPlasmaNlteExcitation.time_prepare_bound_bound_rate_matrix                                  140±1ms
```

---

What are the main changes?

In the file `benchmarks/benchmark_base.py`, I created an inner class `CustomPyTestRequest` which is the replacement for the `request` variable for the function `def regression_data(request)` in the file `tardis/tests/fixtures/regression_data.py`.

In the base benchmark, I created an instance of the class `CustomPyTestRequest` and passed the parameters for module name, node name, regression data directory and path.

When some benchmark has parameters (`@parameterize`), it is more dirty because it needs to specify what benchmark is running between parameters. In the demo, I create a condition to detect the ID of the parameter:

```python
@parameterize({'Matrix': [
    {
        "coll_exc_coeff_values": [1, -2, 3],
        "coll_deexc_coeff_values": [4, 9, 10],
        "number_of_levels": 3,
    },
    {
        "coll_exc_coeff_values": [0.21, 0.045, 0.1234],
        "coll_deexc_coeff_values": [0.7865, 0.987, 0.00123],
        "number_of_levels": 3,
    },
]})
def time_coll_exc_deexc_matrix(self, matrix):
    coll_exc_coeff_values = matrix["coll_exc_coeff_values"]
    coll_deexc_coeff_values = matrix["coll_deexc_coeff_values"]
    number_of_levels = matrix["number_of_levels"]
```

```python
test_parameter_id = 0
if coll_exc_coeff_values != [1, -2, 3]:
    test_parameter_id = 1
```
Then, replace in the node name:
```python
custom_request = self.CustomPyTestRequest(
    f"test_coll_exc_deexc_matrix__coll_exc_coeff_values{test_parameter_id}-coll_deexc_coeff_values{test_parameter_id}-3__",
    …
```

Reply:
Can you put this comment into a github discussion to preserve it?